src/mpl/include/mpl_atomic_opa.h

```diff
@@ -0,0 +1,218 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef MPL_ATOMIC_OPA_H_INCLUDED
+#define MPL_ATOMIC_OPA_H_INCLUDED
+
+#include <stdint.h>
+#include "../../openpa/src/opa_primitives.h"
+
+/* int, int32_t, and uint32_t */
+
+struct MPL_atomic_int32_static_assert_t {
+    /* The following causes an error if the assumption does not hold. */
+    int assert1[sizeof(int) == 4 ? 1 : -1];
```

Comment:
What are the reasons for these asserts? `int` doesn't have to be 32-bit, and `int32_t` is guaranteed to be 32-bit by standard, right?

Reply:
I see. It is because openpa only provide `int` type. Only wrap `int` then, and raise error for `int32_t` and `int64_t`. Right now we don't use `int32_t` or `int64_t` in mpich. If we use it, we'll simply guard against the fallback to opa, which is no big deal.

Reply:
As far as I remember, we are using one of `int32_t`, `uint32_t`, `uint64_t`, and `uint64_t`. This should be one of the motivations of MPL/atomics.

This is compilable, so I would like to leave it as it is for completeness.

Reply:
What happens when `sizeof(int)` is `8`?

Reply:
`int array[-1]` causes an compilation error, which is better than execution-time error and is easier to write than configuration-time error. This static assertion technique is used in, for example, Linux kernel. http://lxr.linux.no/linux+v2.6.26.5/include/linux/kernel.h#L494

Reply:
What is your solution on system with 64-bit `int`s?

Reply:
It means that OPA only supports 64 bit atomicity (assuming a pointer size is 64 bits). We can use 64-bit operations for 32 bits int if needed, but I don't think it practically happens. The same thing goes to 1-2 bytes `int`, which is allowed in terms of specification.

Reply:
> assuming a pointer size is 64 bits

That is obviously a bad assumption, right?

Reply:
Oh my sentence was misleading; if both `void *` and `int` are 8 bytes, it means OPA only supports 64 bit atomicity.

Surely there might be a system that has 4 byte `void *` and 8 byte `int`. Do we need to handle this corner case?

Reply:
What about both types are 4 bytes?

Reply:
It is impossibleto support 8 byte atomic operations with OPA.

Reply:
> 
> 
> It is impossibleto support 8 byte atomic operations with OPA.

That is ok :) On 32-bit system, there is no need for you to support 8-byte atomics.